Orleans.Consensus/Log/StreamLog.cs

```diff
@@ -0,0 +1,191 @@
+        }
+
+        void Rewind(long index)
+        {
+            var result = SetStreamPosition(index);
+            if (!result) throw new ArgumentOutOfRangeException("cannot rewind stream to earlier position");
+            
+            // purge entries from the btree
+            for (var i = index; i <= this.LastLogEntryId.Index; i++)
+            {
+                bTreeIndex.Delete(i);
+                if (this.cache.ContainsKey(i)) this.cache.Remove(i);
+            }
+
+            // warning: this method does not update the LastLogEntryId
+        }
+
+        bool SetStreamPosition(long index)
+        {
+            var indexEntry = this.bTreeIndex.Search(index);
+            if (indexEntry != null)
+            {
+                stream.Position = indexEntry.Pointer;
+                return true;
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// does an existing entry have a different term?
+        /// </summary>
+        /// <param name="entryId"></param>
+        /// <returns></returns>
+        public bool ConflictsWith(LogEntryId entryId)
+        {
+            if (null == entryId) throw new ArgumentNullException(nameof(entryId));
+
+            // entry doesn't exist yet, so it doesn't conflict
+            if (entryId.Index > this.LastLogEntryId.Index || entryId.Index < this.FirstLogEntryId.Index) return false;
+
+            var entry = Get(entryId.Index);
+            
+            if (entry.Id.Term != entryId.Term) return true;
+            return false;
+        }
+
+        public bool Contains(LogEntryId entryId)
+        {
+            if (null == entryId) throw new ArgumentNullException(nameof(entryId));
+
+            if (entryId > this.LastLogEntryId || entryId < this.FirstLogEntryId) return false;
+
+            return this.Get(entryId.Index).Id == entryId;
+        }
+
+        public LogEntry<TOperation> Get(long index)
+        {
+            if (index > this.LastLogEntryId.Index || index < this.FirstLogEntryId.Index) throw new ArgumentOutOfRangeException("outside bounds of index");
+
+            if (this.cache.ContainsKey(index)) return this.cache[index];
+
+            var indexEntry = this.bTreeIndex.Search(index);
+            if (null == indexEntry) throw new ArgumentOutOfRangeException("cannot find index");
+
+            stream.Position = indexEntry.Pointer;
+            return serializer.Deserialize(stream).Result;
+        }
+
+        public IEnumerable<LogEntry<TOperation>> GetCursor(long fromIndex)
+        {
+            for (var i = fromIndex; i <= this.LastLogEntryId.Index; i++)
+            {
+                yield return Get(i);
+            }
+        }
+
+        public IEnumerable<LogEntry<TOperation>> GetReverseCursor()
+        {
+            if (this.stream.Length == 0) yield break;
+
+            for (var i = this.LastLogEntryId.Index; i >= this.FirstLogEntryId.Index; i--)
+            {
+                yield return Get(i);
```

Comment:
If we use this a lot, maybe we should consider creating some kind of `ReverseEnumerator` class on the tree.